### PR TITLE
Add NumPy-like properties to GridType and length to StructureType

### PR DESCRIPTION
**Pull request by laliberte**
This adresses #43. Tests are included and everything appears to be working fine.

### CODE REVIEW

**Comment by laliberte** [comment]
@jameshiebert do you give your thumbs up? This one seems pretty straightforward.

**Comment by shoyer** [comment]
It's worth double checking, but I think this is all we need for xarray's pydap integration to work again.

**Comment by laliberte** [comment]
@shoyer, I'm not going to close the issue right away. If it's not sufficient, then we'll look at what's required. +1 for merging?

File: src/pydap/model.py [hunk 1]
```diff
@@ -333,6 +333,9 @@ def keys(self):
         """Method to emulate a dictionary, returning keys."""
         return self._keys[:]
 
+    def __len__(self):
```
Comment (shoyer) [inline src/pydap/model.py]: Consider making `StructureType` inherit from `collections.Mapping` as well. That guarantees you don't forget anything, and helps for type checks (e.g., with `isinstance` or `typing.Mapping`).

File: src/pydap/model.py [hunk 2]
```diff
@@ -568,6 +571,20 @@ def __getitem__(self, key):
...
+
+    @property
+    def dtype(self):
+        """Return the first children dtype."""
+        return self.array.dtype
+
+    @property
+    def shape(self):
+        """Return the first children shape."""
+        return self.array.shape
```
Comment (shoyer) [inline src/pydap/model.py]: It's not strictly necessary, but I usually make my duck array types implement `__array__`, `ndim` and `size` as well:
https://github.com/pydata/xarray/blob/371d034372bc7522098a142a0debf93916c49102/xarray/core/utils.py#L406

(Implement these on `BaseType`, too, if you decide to go this route)

File: src/pydap/model.py (hunk 1 above)
Comment (laliberte) [inline src/pydap/model.py]: done

File: src/pydap/model.py (hunk 2 above)
Comment (laliberte) [inline src/pydap/model.py]: And done!

File: src/pydap/model.py (hunk 1 above)
Comment (laliberte) [inline src/pydap/model.py]: I've made it into a ``MutableMapping`` because this is what it is. Sounds good to you?

**Comment by laliberte** [comment]
@shoyer could you confirm that ``MutableMapping`` is as good (if not better than) as ``Mapping``? If this is all good, anyone has anything else to say before I merge?

**Comment by shoyer** [comment]
> @shoyer could you confirm that MutableMapping is as good (if not better than) as Mapping? If this is all good, anyone has anything else to say before I merge?

Be a little careful with MutableMapping -- it adds quite a few methods like pop, popitem, clear, update, and setdefault. It's probably fine in this case, but it's not desirable for every mutable Mapping type (e.g., xarray.Dataset is a collections.Mapping instance, but not a collections.MutableMapping).

**Comment by laliberte** [comment]
Ok, that's a good point. Let's be on the safe side and use Mapping.

**Pull request merged**